### ISSUE TITLE
gemma 4 performance reduced after upgrading to omlx v0.6.1

### ISSUE THREAD

**Issue by noonkho** [jundot/omlx#2786]
**Describe the bug**
Model: mlx-community/gemma-4-31B-it-qat-mxfp8 (Pin and Loaded)
with VLM MTP: mlx-community/gemma-4-31B-it-qat-assistant-mxfp8

Before this version, gemma 4 is outputting 30+ token per second.
Now it's 16.4 tok/s.

Also realize the hit rate for gemma 4 is like terrible. I wonder if there are better gemma 4 models or config to setup in omlx?

**To Reproduce**
1. Download Gemma 4 model
2. Load and Run Gemma 4 model
3. Run the model and look at the dashboard between previous version and the current version

**Expected behavior**
Performance dropped

**Desktop (please complete the following information):**
 - macOS Version: [e.g. 26.4.1]
 - oMLX Version: v0.6.1

**Comment by jundot** [comment]
Thanks for the report. I think this is from #2752 in v0.6.1. It turns off the MTP drafter for a request when any other request is active at the same time, and that request then decodes without MTP until it finishes. So background requests from your chat UI (like title generation) or overlapping turns can keep gemma 4 stuck at plain decode speed, and 16.4 tok/s matches that baseline.

Could you switch the log level to debug in global settings and attach the server log here? If it shows "vlm_mtp routing skipped" lines that confirms it, and I'll work on improving the gating.

The hit rate is a separate thing, gemma's sliding window layers limit prefix cache reuse, so a low rate is expected regardless of version.

**Comment by noonkho** [comment]
I just updated to v0.6.2 and the performance is back to before which is 30+ token/second with the VLM MTP (tested in the chat UI).

The server log is as follows (should be normal):
2026-08-19 12:53:39,749 - omlx.engine_pool - INFO - [-] - Preloading pinned model: gemma-4-31B-it-qat-mxfp8
2026-08-19 12:53:39,749 - omlx.engine_pool - INFO - [-] - Loading model: gemma-4-31B-it-qat-mxfp8
2026-08-19 12:53:46,307 - omlx.cache.vision_feature_cache - INFO - [-] - Vision feature SSD cache scan: scanned=424, indexed=424, errors=0, total_size=6374.0MB
2026-08-19 12:53:46,307 - omlx.engine.vlm - INFO - [-] - Vision feature cache enabled (SSD: /Users/echingkho/.omlx/cache/vision_features)
2026-08-19 12:53:48,304 - omlx.scheduler - INFO - [-] - Loaded 2 additional EOS token(s) from generation_config.json: {106, 50}
2026-08-19 12:53:48,304 - omlx.scheduler - INFO - [-] - Aligning paged cache block_size=256 to 1024 (RotatingKVCache window_size=1024, multiplier=1x)
2026-08-19 12:53:48,304 - omlx.memory_monitor - INFO - [-] - MemoryMonitor initialized (estimator-only, eviction disabled)
2026-08-19 12:53:48,304 - omlx.memory_monitor - INFO - [-] - Model info set: 60 layers (10 KVCache, rotating 50x@1024), 16 KV heads, 32 Q heads, 256 head_dim. Estimated memory per 64-token block: 60.00 MB
2026-08-19 12:53:48,304 - omlx.scheduler - INFO - [-] - paged SSD-only mode: max_blocks=100000, block_size=1024 tokens
2026-08-19 12:53:48,304 - omlx.cache.paged_cache - INFO - [-] - PagedCacheManager initialized: block_size=1024, initial_blocks=1024, max_blocks=100000, max_tokens=102400000
2026-08-19 12:53:48,305 - omlx.cache.paged_ssd_cache - INFO - [-] - Scanning SSD cache directory: /Users/echingkho/.omlx/cache
2026-08-19 12:53:48,401 - omlx.cache.paged_ssd_cache - INFO - [-] - SSD cache scan complete: scanned=109, indexed=0, errors=0, total_size=2.01 GB, gdn_sidecars=0, skipped_incompatible=108 blocks (2.01 GB)
2026-08-19 12:53:48,402 - omlx.cache.paged_ssd_cache - INFO - [-] - PagedSSDCacheManager initialized: dir=/Users/echingkho/.omlx/cache, max_size=744.00 GB, hot_cache=61.00 GB, existing_files=0, disk_free=2618.85 GB, cache_used=2.01 GB, incompatible_files=108
2026-08-19 12:53:48,402 - omlx.cache.paged_cache - INFO - [-] - paged SSD cache manager connected to PagedCacheManager
2026-08-19 12:53:48,402 - omlx.cache.prefix_cache - INFO - [-] - PagedSSDCacheManager connected to BlockAwarePrefixCache
2026-08-19 12:53:48,402 - omlx.scheduler - INFO - [-] - paged SSD cache enabled: cache_dir=/Users/echingkho/.omlx/cache, max_size=744.00 GB, block_size=1024 tokens
2026-08-19 12:53:48,402 - omlx.scheduler - INFO - [-] - paged SSD cache enabled: /Users/echingkho/.omlx/cache, block_size=1024, max_blocks=100000
2026-08-19 12:53:48,402 - omlx.scheduler - INFO - [-] - Output parser detected: gemma4 for gemma-4-31B-it-qat-mxfp8, stop_tokens=[]
2026-08-19 12:53:48,402 - omlx.engine_core - INFO - [-] - Engine started
2026-08-19 12:53:48,403 - omlx.cache.paged_ssd_cache - INFO - [-] - PagedSSDCacheManager updated layer cache signature (60 layers, 2 unique types, turboquant_kv_bits=None, cachelist_subtypes=no)
2026-08-19 12:53:48,448 - omlx.engine.vlm - INFO - [-] - VLM tool calling enabled: parser=gemma4
2026-08-19 12:53:48,451 - omlx.engine.vlm - INFO - [-] - VLMBatchedEngine loaded: /Users/echingkho/.omlx/models/mlx-community/gemma-4-31B-it-qat-mxfp8
2026-08-19 12:53:48,451 - mlx_vlm.speculative.drafters - INFO - [-] - Auto-detected --draft-kind='mtp' for drafter '/Users/echingkho/.omlx/models/mlx-community/gemma-4-31B-it-qat-assistant-mxfp8' (model_type='gemma4_assistant').
2026-08-19 12:53:49,844 - omlx.speculative.vlm_mtp - INFO - [-] - VLM MTP drafter loaded: path=/Users/echingkho/.omlx/models/mlx-community/gemma-4-31B-it-qat-assistant-mxfp8 kind=mtp model_type=gemma4_assistant
2026-08-19 12:53:49,845 - omlx.scheduler - INFO - [-] - VLM MTP drafter attached to scheduler (block_size=6)
2026-08-19 12:53:49,845 - omlx.engine.vlm - INFO - [-] - VLM MTP drafter attached to engine: /Users/echingkho/.omlx/models/mlx-community/gemma-4-31B-it-qat-mxfp8 (block_size=6)
2026-08-19 12:53:49,845 - omlx.engine_pool - INFO - [-] - VLM MTP enabled for gemma-4-31B-it-qat-mxfp8, drafter=gemma-4-31B-it-qat-assistant-mxfp8
2026-08-19 12:53:49,847 - omlx.engine_pool - INFO - [-] - Loaded model: gemma-4-31B-it-qat-mxfp8 (actual: 32.31GB, local estimate: 32.71GB, full model: 32.71GB, total: 204.06GB)
2026-08-19 12:57:03,498 - httpx - INFO - [-] - HTTP Request: GET https://huggingface.co/api/models?author=mlx-community&sort=trendingScore&limit=60&expand=safetensors&expand=downloads&expand=likes&expand=trendingScore "HTTP/1.1 200 OK"
2026-08-19 12:57:03,500 - httpx - INFO - [-] - HTTP Request: GET https://huggingface.co/api/models?author=mlx-community&sort=downloads&limit=60&expand=safetensors&expand=downloads&expand=likes&expand=trendingScore "HTTP/1.1 200 OK"
2026-08-19 12:57:03,501 - httpx - INFO - [-] - HTTP Request: GET https://huggingface.co/api/models?author=mlx-community&sort=downloads&limit=60&expand=safetensors&expand=downloads&expand=likes&expand=trendingScore "HTTP/1.1 200 OK"
2026-08-19 12:57:03,513 - httpx - INFO - [-] - HTTP Request: GET https://huggingface.co/api/models?author=mlx-community&sort=trendingScore&limit=60&expand=safetensors&expand=downloads&expand=likes&expand=trendingScore "HTTP/1.1 200 OK"
2026-08-19 12:58:56,387 - omlx.engine.vlm - INFO - [-] - GrammarCompiler initialized for /Users/echingkho/.omlx/models/mlx-community/gemma-4-31B-it-qat-mxfp8
2026-08-19 12:58:57,994 - omlx.scheduler - INFO - [-] - vlm_mtp decode started: request=48ef3b48-2f4d-44ff-bff5-fa0019010987 uid=-1 block_size=6
2026-08-19 12:59:28,190 - omlx.scheduler - INFO - [-] - Enabled boundary cache snapshots for stateful non-sliceable cache layers
2026-08-19 12:59:44,684 - omlx.scheduler - INFO - [-] - vlm_mtp stats: request=48ef3b48-2f4d-44ff-bff5-fa0019010987 finish=stop rounds=555 accepted=956/2775 (34.5%) tokens_per_round=2.72 emitted=1510 block_size=6
2026-08-19 12:59:44,691 - omlx.scheduler - INFO - [-] - Cache phase timings: boundary_capture_extract=0.0ms/1, boundary_capture_sync=0.0ms/1, cleanup_finished_sync=0.0ms/1, store_cache_main_boundary=0.0ms/1
2026-08-19 12:59:44,693 - omlx.server - INFO - [-] - Chat completion: model=gemma-4-31B-it-qat-mxfp8, 1509 tokens in 48.28s (32.9 tok/s), prompt: 34, finish_reason=stop, max_tokens=16384, request_max_tokens=None

**Comment by noonkho** [comment]
Actually when I tried to call it with OpenAI API, it skips the vlm_mtp routing. Still dropped performance  😭.

2026-08-19 13:06:58,305 - omlx.scheduler - INFO - [-] - vlm_mtp routing skipped for 907b89d7-d0ff-4f07-8911-f77e4c5e9a25: request carries per-request logits processors without vlm_mtp support (GrammarConstraintProcessor); falling back to BatchGenerator
2026-08-19 13:07:17,196 - omlx.scheduler - INFO - [-] - Using boundary cache snapshot for 907b89d7-d0ff-4f07-8911-f77e4c5e9a25: storing 3072/3738 tokens (skipping trailing partial block, 2 intermediate snapshots)
2026-08-19 13:07:17,226 - omlx.scheduler - INFO - [-] - Cache phase timings: boundary_capture_extract=0.0ms/1, boundary_capture_sync=0.0ms/1, cleanup_finished_sync=5.4ms/2, store_cache_main_boundary=2.8ms/2, store_cache_main_collect=0.0ms/1, store_cache_main_dispatch=25.5ms/1, store_cache_worker_sync=0.2ms/1
2026-08-19 13:07:17,229 - omlx.server - INFO - [-] - Chat completion: model=gemma-4-31B-it-qat-mxfp8, 317 tokens in 30.38s (17.2 tok/s), prompt: 3421, finish_reason=stop, max_tokens=4096, request_max_tokens=4096
2026-08-19 13:07:23,894 - omlx.scheduler - INFO - [-] - vlm_mtp routing skipped for 7778867e-9781-44ce-b227-5097a807bd60: request carries per-request logits processors without vlm_mtp support (GrammarConstraintProcessor); falling back to BatchGenerator
2026-08-19 13:08:23,221 - omlx.scheduler - INFO - [-] - Using boundary cache snapshot for 7778867e-9781-44ce-b227-5097a807bd60: storing 4096/5047 tokens (skipping trailing partial block, 1 intermediate snapshots)
2026-08-19 13:08:23,250 - omlx.scheduler - INFO - [-] - Cache phase timings: boundary_capture_extract=0.0ms/2, boundary_capture_sync=6.1ms/2, boundary_snapshot_save=117.8ms/1, cleanup_finished_sync=10.9ms/3, store_cache_main_boundary=3.4ms/3, store_cache_main_collect=0.0ms/2, store_cache_main_dispatch=50.9ms/2, store_cache_worker_sync=0.2ms/1
2026-08-19 13:08:23,252 - omlx.server - INFO - [-] - Chat completion: model=gemma-4-31B-it-qat-mxfp8, 995 tokens in 65.95s (16.9 tok/s), prompt: 4052, finish_reason=stop, max_tokens=4096, request_max_tokens=4096
2026-08-19 13:08:27,360 - omlx.scheduler - INFO - [-] - vlm_mtp routing skipped for fbadb360-76cb-44ef-ad5f-ae08b6f52ae5: request carries per-request logits processors without vlm_mtp support (GrammarConstraintProcessor); falling back to BatchGenerator
2026-08-19 13:08:38,448 - omlx.scheduler - INFO - [-] - Using boundary cache snapshot for fbadb360-76cb-44ef-ad5f-ae08b6f52ae5: storing 3072/3402 tokens (skipping trailing partial block, 0 intermediate snapshots)
2026-08-19 13:08:38,478 - omlx.scheduler - INFO - [-] - Cache phase timings: boundary_capture_extract=0.0ms/2, boundary_capture_sync=6.1ms/2, boundary_snapshot_save=117.8ms/1, cleanup_finished_sync=16.2ms/4, store_cache_main_boundary=6.2ms/4, store_cache_main_collect=0.0ms/3, store_cache_main_dispatch=76.3ms/3, store_cache_worker_sync=0.3ms/2
2026-08-19 13:08:38,479 - omlx.server - INFO - [-] - Chat completion: model=gemma-4-31B-it-qat-mxfp8, 184 tokens in 15.15s (17.3 tok/s), prompt: 3218, finish_reason=stop, max_tokens=4096, request_max_tokens=4096
2026-08-19 13:08:38,696 - omlx.scheduler - INFO - [-] - prefix cache: request 608ddfc3-7624-42cf-b9ef-03c30d84d045 re-prefills 4664 of 6712 tokens (reused 2048); closest stored sequence 907b89d7-d0ff-4f07-8911-f77e4c5e9a25 shares the first 2392 of 3072 comparable tokens before diverging
2026-08-19 13:08:53,049 - omlx.scheduler - INFO - [-] - vlm_mtp routing skipped for 608ddfc3-7624-42cf-b9ef-03c30d84d045: request carries per-request logits processors without vlm_mtp support (GrammarConstraintProcessor); falling back to BatchGenerator
